crates/shared/src/http_client.rs

```diff
@@ -19,11 +23,141 @@ pub async fn response_body_with_size_limit(
     Ok(bytes)
 }
 
+#[derive(Debug, Clone)]
+pub struct RateLimitingStrategy {
+    drop_requests_until: Instant,
+    /// How many requests got rate limited in a row.
+    times_rate_limited: u64,
+    back_off_growth_factor: f64,
+    min_back_off: Duration,
+    max_back_off: Duration,
+}
+
+impl RateLimitingStrategy {
+    pub fn try_new(
+        back_off_growth_factor: f64,
+        min_back_off: Duration,
+        max_back_off: Duration,
+    ) -> Result<Self> {
+        ensure!(
+            back_off_growth_factor.is_normal(),
+            "back_off_growth_factor must be a normal f64"
+        );
+        ensure!(
+            back_off_growth_factor > 1.0,
+            "back_off_growth_factor needs to be greater than 1.0"
+        );
+        ensure!(
+            min_back_off <= max_back_off,
+            "min_back_off needs to be <= max_back_off"
+        );
+        Ok(Self {
+            drop_requests_until: Instant::now(),
+            times_rate_limited: 0,
+            back_off_growth_factor,
+            min_back_off,
+            max_back_off,
+        })
+    }
+}
+
+impl RateLimitingStrategy {
+    /// Resets back off and stops rate limiting requests.
+    pub fn response_ok(&mut self) {
+        self.times_rate_limited = 0;
+        self.drop_requests_until = Instant::now();
+    }
+
+    /// Calculates back off based on how often we got rate limited in a row.
+    fn get_current_back_off(&self) -> Duration {
+        let mut back_off = self.min_back_off;
+        for _ in 0..self.times_rate_limited {
```

Comment:
This should use f64::powi or f64::powf instead of a loop. You can't get overflow panics with f64 because worst case it will stay at +INF.

Reply:
I wanted to avoid an overflow with [Duration::mul_f64()](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.mul_f64).
Being able to compute the correct factor to call `mul_f64()` only once wouldn't help us much if that would cause `mul_f64()` to panic because the result would overflow `Duration`.

Reply:
is there a `checked_mul_f64`?

Reply:
Could be like
```rust
let factor =  back_off_growth_factor.pow(times_rate_limited);
back_off *= factor;
```
I see now that Duration for some reason doesn't have a non panicking version of mul_f64 which is weird. In this case I feel it is reasonable to extract the f64 out of the duration, do the math, put it back.

Reply:
Unfortunately there is only [checked_mul](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.checked_mul) and [saturating_mul](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.saturating_mul) which both require a u32 as the argument.
I could make the `growth_factor` argument an `u32` but that felt overly restrictive.
But I don't feel strongly about that so I could change it if you'd like that.

Reply:
> In this case I feel it is reasonable to extract the f64 out of the duration, do the math, put it back.

I hadn't considered that before. Good idea. 👍